pilot/pkg/xds/ep_filters.go

```diff
@@ -111,10 +112,16 @@ func EndpointsByNetworkFilter(push *model.PushContext, proxyNetwork string, endp
 				}
 				// TODO: figure out a way to extract locality data from the gateway public endpoints in meshNetworks
 				gwEp.Metadata = util.BuildLbEndpointMetadata(network, model.IstioMutualTLSModeLabel)
-				lbEndpoints = append(lbEndpoints, gwEp)
+				remoteLbEps = append(remoteLbEps, gwEp)
 			}
 		}
 
+		// Only include remote endpoints for a gateway if we have no local endpoints
+		// Normal proxies always get all the endpoints.
+		if proxyType != model.Router || (len(remoteLbEps) > 0 && len(lbEndpoints) == 0) {
```

Comment:
Could this egress gateway always send traffic to local cluster endpoints?

Reply:
I understand for a gateway, only add the endpoints of remote cluster/network if there is no  local endpoints

Reply:
@stevenctl

Reply:
Egress gateways should be sending traffic outside the mesh. This is intended to force local network endpoints, not local cluster. We can't expect every cluster in a network to run an ingress.

Reply:
But this is a regression in some degree, lets discuss it. @howardjohn

Reply:
@stevenctl are we only filtering other gateways? IE if I have my gateway in network "A" and I am sending a request to some external service in network "B" (not a gateway!), will it still show up?

Reply:
If we're building endpoints for a gateway, we will not send traffic to another gateway.

In what case would we set network for an external service? IIUC the primary purpose of setting network is to intentionally go through an istio gateway with passthrough.

Reply:
So if somehow we had things setup to go pod -> egress gw (net A) -> ingress (net B) -> svc net B, that would get blocked. 

pod -> egress -> external service should be fine, not sure where networks come into play though.

Reply:
Actually we did support clients in a cluster to send traffic out via egress gateway,  in this case 

client pods in cluster A  --> egress gateway in cluster A  ---> server pods in cluster A
                                                               | 
                                                               |   X
                                                               | 
                                                              cluster B pods

Here one cluster is in one network

Reply:
Ah. I suppose in that case we're not at risk of the bug this intends to fix, and we should still be able to go cross-network. I'll look into how to determine this is only for _ingress_ gateway. Thanks!

Reply:
@hzxuzhonghu I am curious what is the usecase for client pods to go via egress gateway to server pods in same cluster and network?

Reply:
Just when the server instances are across networks and let's say the client can not access the remote network gateway.

Reply:
From the client's view, they are not aware of where the server resides. Only a service hostname is known

Reply:
Ok. Only when server instances span across multiple networks?

Reply:
Yes. Rare use case

Reply:
Ok. got it. thanks